Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/shape_smoothing.h

```diff
@@ -0,0 +1,415 @@
+*  \cgalParamBegin{vertex_point_map} the property map with the points associated
+*    to the vertices of `pmesh`. Instance of a class model of `ReadWritePropertyMap`.
+*  \cgalParamEnd
+*  \cgalParamBegin{vertex_is_constrained_map} a property map containing the
+*    constrained-or-not status of each vertex of `pmesh`. A constrained vertex
+*    cannot be modified at all during smoothing.
+*  \cgalParamEnd
+*  \cgalParamBegin{number_of_iterations} the number of iterations for the
+*    sequence of the smoothing iterations performed. Each iteration is performed
+*    with the given time step.
+*  \cgalParamEnd
+*  \cgalParamBegin{use_explicit_scheme} a normalized explicit scheme for smoothing along the curvature flow.
+*    It is not dependent on the time step parameter. Each vertex is moved sequentially without solving
+*    a linear system. It can be useful for subtle noise removal and to avoid the result
+*    to converge to a sphere.
+*  \cgalParamEnd
+* \cgalNamedParamsEnd
+*/
+template<typename PolygonMesh, typename FaceRange, typename NamedParameters>
+void smooth_along_curvature_flow(const FaceRange& faces, PolygonMesh& pmesh, const double& time,
+                                 const NamedParameters& np)
+{
+  if (boost::begin(faces)==boost::end(faces))
+    return;
+
+  using boost::choose_param;
+  using boost::get_param;
+
+  #ifdef CGAL_PMP_SMOOTHING_VERBOSE
+  CGAL::Timer t;
+  std::cout << "Initializing...";
+  std::cout.flush();
+  t.start();
+  #endif
+
+  typedef typename GetGeomTraits<PolygonMesh, NamedParameters>::type GeomTraits;
+
+  typedef typename GetVertexPointMap<PolygonMesh, NamedParameters>::type VertexPointMap;
+  VertexPointMap vpmap = choose_param(get_param(np, internal_np::vertex_point),
+                               get_property_map(CGAL::vertex_point, pmesh));
+
+  typedef typename boost::graph_traits<PolygonMesh>::vertex_descriptor vertex_descriptor;
+  typedef typename boost::lookup_named_param_def <
+      internal_np::vertex_is_constrained_t,
+      NamedParameters,
+      internal::Constrained_vertices_map<vertex_descriptor>
+    > ::type VCMap;
+  VCMap vcmap = choose_param(get_param(np, internal_np::vertex_is_constrained),
+                             internal::Constrained_vertices_map<vertex_descriptor>());
+
+  std::size_t nb_iterations = choose_param(get_param(np, internal_np::number_of_iterations), 1);
+
+  bool use_explicit_scheme = choose_param(get_param(np, internal_np::use_explicit_scheme), false);
+
+  if(use_explicit_scheme)
+  {
+    smooth_curvature_flow_explicit(faces, pmesh, parameters::number_of_iterations(nb_iterations));
+  }
+  else
+  {
+    // implicit scheme
+#if defined(CGAL_EIGEN3_ENABLED)
+  #if EIGEN_VERSION_AT_LEAST(3,2,0)
+    typedef typename Eigen::SparseMatrix<double> Eigen_sparse_matrix;
+    typedef typename Eigen::BiCGSTAB<Eigen_sparse_matrix, Eigen::IncompleteLUT<double> > Eigen_solver;
+    typedef CGAL::Eigen_solver_traits<Eigen_solver> Default_solver;
+  #else
```

Comment:
The solver must be a template parameter! You can use a new named parameter for if with Eigen as default and if eigen is not available a static assertion.